ndonnx/_typed_array/indexing.py

```diff
@@ -0,0 +1,269 @@
+    ):
+        key = onnx.const(np.array(key), dtype=onnx.bool_)
+
+    # Fish out the boolean masks before normalizing to tuples
+    if isinstance(key, onnx.TyArrayBool):
+        return key
+
+    if isinstance(key, int | slice | EllipsisType | None | onnx.TyArrayInt64):
+        return (_normalize_getitem_key_item(key),)
+
+    if isinstance(key, tuple):
+        # Cannot use `count` since that will trigger an __eq__ call
+        # with type promotion against `...`
+        if sum(isinstance(el, type(...)) for el in key) > 1:
+            raise IndexError("more than one ellipsis (`...`) in index tuple")
+
+        return tuple(_normalize_getitem_key_item(el) for el in key)
+
+    raise IndexError(f"unexpected key `{key}`")
+
+
+def _normalize_getitem_key_item(key: GetitemItem) -> GetitemItem:
+    if isinstance(key, slice):
+        slice_kwargs = {"start": key.start, "stop": key.stop, "step": key.step}
+        out: dict[str, int | None | onnx.TyArrayInt64] = {}
+        for k, arg in slice_kwargs.items():
+            if not isinstance(arg, None | int | onnx.TyArrayInt64):
+                raise IndexError(f"slice argument `{k}` has invalid type `{type(arg)}`")
+            if isinstance(arg, onnx.TyArrayBase):
+                if arg.dtype not in (onnx.int32, onnx.int64) or arg.ndim != 0:
+                    raise IndexError(
+                        f"array-slice argument must be be an int64 scalar. Found `{arg}`"
+                    )
+                out[k] = arg.astype(onnx.int64)
+            else:
+                out[k] = arg
+        return slice(*out.values())  # Beware that we use the dict order here!
+
+    if isinstance(key, int | None | EllipsisType):
+        return key
+
+    if key.ndim != 0:
+        raise IndexError(
+            f"index arrays must be rank-0 tensors; found rank `{key.ndim}`"
+        )
+    return key.astype(onnx.int64)
+
+
+def key_to_indices(key: tuple[slice | int, ...], shape: TyArrayInt64) -> TyArrayInt64:
+    """Compute expanded indices for ``key`` for an array of shape ``shape``."""
+    # TODO: Allow dynamic inputs to DType.__ndx_arange__?
+    import ndonnx as ndx
+
+    indices_per_dim = []
+    for s, length in zip(key, shape):
+        if isinstance(s, int):
+            stop = None if s == -1 else s + 1
+            s = slice(s, stop, 1)
+        indices_per_dim.append(
+            ndx.asarray(ort_compat.range(*[el._var for el in _get_indices(s, length)]))
+        )
+
+    grid = ndx.meshgrid(*indices_per_dim, indexing="ij")
+    indices = ndx.concat([ndx.reshape(el, (-1, 1)) for el in grid], axis=-1)
+
+    return safe_cast(onnx.TyArrayInt64, indices._tyarray)
+
+
+def _get_indices(
+    s: slice, length: int | TyArrayInt64
+) -> tuple[TyArrayInt64, TyArrayInt64, TyArrayInt64]:
+    """Array-compatible implementation of ``slice.indices``.
+
+    Slice members must be of type ``int`` or ``None`` while `length` may be an array.
+    """
+    from .funcs import maximum, minimum
```

Comment:
I believe this lazy import can go by either directly using the Vars (we know the dtype is i64) or by using the dunder members.